File: LinkourmetALS.py
import sys
# import itertools
# from math import sqrt
# from operator import add
# from os.path import join, isfile, dirname

from pyspark import SparkConf, SparkContext
from pyspark.mllib.recommendation import ALS, Rating

def init():
    conf = SparkConf().setAppName("LinkourmetALS")
    return SparkContext(conf=conf)


def loadData(sc):
    # currently load from a file, TODO: load from database
    return sc.textFile("ratings.dat").map(lambda l: l.split('::')).map(lambda l: (int(l[0]), int(l[1]), float(l[2])))


def getUserData(data, userID):
    sharedLinks = data.filter(lambda r: r[0] == userID).map(lambda r: r[1])
    allLinks = data.map(lambda r: r[1]).distinct()

    unsharedLinks = allLinks.subtract(sharedLinks)
    return unsharedLinks.map(lambda x: (userID, x))


def findBestModel(data):
    # Build the recommendation model using Alternating Least Squares
    # TODO need to test the best configuration of model
    rank = 10
    numIterations = 20
    return ALS.trainImplicit(data, 1)


def recommend(model, userData):
    predictions = model.predictAll(userData).map(lambda x: (x[0], x[1], x[2])).collect()
    recommendations = sorted(predictions, key=lambda x: x[2], reverse=True)
    return recommendations


if __name__ == "__main__":
    
    sc = init()

    if len(sys.argv) != 2:
        sys.exit('Usage: [spark root]/bin/spark-submit %s {userID}' % sys.argv[0])
    
    userID = int(sys.argv[1])

    data = loadData(sc) # data is RDD of (user, link, rating)
    print('[XC] All Data:')
    print(data.collect());

    model = findBestModel(data)

    userData = getUserData(data, userID) # userData is RDD of (user, link) where 'link' is not shared by this user
    print('[XC] User Data:')
    print(userData.collect());

    reclinks = recommend(model, userData)
    print('[XC] Recommend links:')
    for linkInfo in reclinks:
        print (linkInfo[1])

    sc.stop()

File: LinkourmetALS.scala
import java.io.File

import scala.io.Source

import org.apache.log4j.Logger
import org.apache.log4j.Level

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.rdd._
import org.apache.spark.mllib.recommendation.{ALS, Rating, MatrixFactorizationModel}

object LinkourmetALS {

  def main(args: Array[String]) {

    Logger.getLogger("org.apache.spark").setLevel(Level.WARN)

    if (args.length != 1) {
      println("Usage: YOUR_SPARK_HOME/bin/spark-submit --class LinkourmetALS target/scala-*/*.jar userID")
      sys.exit(1)
    }

    // set up environment

    val conf = new SparkConf().setAppName("LinkourmetALS")
    val sc = new SparkContext(conf)
    
    // load data

    val data = sc.textFile("ratings.dat").map(_.split("::") match { case Array(user, link, rate) =>
        Rating(user.toInt, link.toInt, rate.toDouble)
      })

    println("[XC] Data: ")
    data.collect().foreach(println)

    // Build the recommendation model using ALS

    val rank = 10
    val numIterations = 20
    val model = ALS.trainImplicit(data, 1, 5)

    // userData is RDD of (user, link) where 'link' is not shared by this user

    val userID = args(0).toInt

    val sharedLinks = data.filter(x => x.user == userID).map(_.product)
    val allLinks = data.map(_.product).distinct()
    val unsharedLinks = allLinks.subtract(sharedLinks)

    val userData = unsharedLinks.map((userID, _))

    println("[XC] User Data: ")
    userData.collect().foreach(println)

    // recommendation

    val predictions = model.predict(userData).collect().sortBy(-_.rating).take(50)
  
    println("[XC] Recommendation: ")
    predictions.foreach(println)

    // clean up
    sc.stop()
  }
} 
